Modules/trigger.py

```diff
@@ -32,17 +34,32 @@ def __init__(self, bot: pydle.BasicClient, words: List[str], words_eol: List[str
                 nickname.
             identified (bool): Whether or not the user is identified with NickServ.
         """
-        self.bot = bot
-        self.words = words
-        self.words_eol = words_eol
-        self.nickname = nickname
-        self.target = target
-        self.ident = ident
-        self.hostname = hostname
-        self.realname = realname
-        self.away = away
-        self.account = account
-        self.identified = identified
+        self._bot = bot
+        self._words = words
+        self._words_eol = words_eol
+        self._nickname = nickname
+        self._target = target
+        self._ident = ident
+        self._hostname = hostname
+        self._realname = realname
+        self._away = away
+        self._account = account
+        self._identified = identified
+
+        self._hash = None
+        self._immutable = True
+
+    bot = property(lambda self: self._bot)
+    words = property(lambda self: self._words)
+    words_eol = property(lambda self: self._words_eol)
+    nickname = property(lambda self: self._nickname)
+    target = property(lambda self: self._target)
+    ident = property(lambda self: self._ident)
+    hostname = property(lambda self: self._hostname)
```

Comment:
Question: does this create a read-only property? we may want to be able to change these values at runtime

Reply:
Omitting the setter makes the property read-only. 
Though technically one could write to the private attribute directly.

Reply:
Of course we could, but that would be bad practice to write to a private directly. What im saying here is we may want to change some of these properties at runtime therefore should not limit them to read-only

Reply:
The Trigger object, as it currently is, is designed to provide a command with information on how and by whom the command was invoked. This should be a snapshot of the state at the point of invocation.
e.g. if a user changes their nickname during execution of a command by them, the nickname in the trigger object should not be updated
The same goes for the other user attributes and even more strongly for words and words_eol.

Reply:
Long story short, the Trigger object is a prime example of an object that really ought to be immutable.

The alternative could lead to bad things when a command function suddenly finds its Trigger object changed in such a way that it no longer matches the state of the command function.

Reply:
Nope, trigger should be immutable, i misread.

Reply:
If my understanding of Trigger objects being invalidated after command execution has finished, write-access after initializing them should be forbidden, due to the fact that information should be reliable  and they are also short-lived, as on the next command a new Trigger would be constructed, reflecting any change that might have happened in the mean time.

I fully agree with MH here

Reply:
YEah. my original comment was based in a misunderstanding on my part, Trigger objects should be immutable after creation, and should not be able to modify the bot or its context.